File: Sources/PetCompanion/Pet/PetStateMachine.swift
import Foundation

enum PetState {
    case walking
    case idle
    case looking
    case watching
    case sleeping
    case stretching
    case meowing
    case paused
}

final class PetStateMachine {
    private(set) var state: PetState = .walking
    private var previousActiveState: PetState = .walking
    private var stateStartedAt: TimeInterval = 0
    private var nextTransitionAt: TimeInterval = 6

    var isPaused: Bool {
        state == .paused
    }

    var isSleeping: Bool {
        state == .sleeping
    }

    func setPaused(_ paused: Bool, at time: TimeInterval) {
        if paused {
            if state != .paused {
                previousActiveState = state
                transition(to: .paused, at: time)
            }
        } else if state == .paused {
            transition(to: previousActiveState, at: time)
        }
    }

    func update(at time: TimeInterval) {
        guard state != .paused, time >= nextTransitionAt else {
            return
        }

        switch state {
        case .walking:
            transition(to: .idle, at: time)
        case .idle:
            transition(to: nextIdleTransition(), at: time)
        case .looking:
            transition(to: .walking, at: time)
        case .watching:
            nextTransitionAt = .infinity
        case .sleeping, .stretching, .meowing:
            transition(to: .walking, at: time)
        case .paused:
            break
        }
    }

    func setWatching(_ watching: Bool, at time: TimeInterval) {
        guard state != .paused else {
            return
        }

        if watching {
            if state == .walking || state == .idle {
                transition(to: .watching, at: time)
            }
        } else if state == .watching {
            transition(to: .walking, at: time)
        }
    }

    func triggerClickReaction(at time: TimeInterval) {
        guard state != .paused else {
            return
        }

        let reaction: PetState = Bool.random() ? .meowing : .stretching
        transition(to: reaction, at: time)
    }

    func triggerSleep(at time: TimeInterval) {
        guard state != .paused else {
            return
        }

        if state == .sleeping {
            transition(to: .walking, at: time)
        } else {
            transition(to: .sleeping, at: time)
        }
    }

    private func nextIdleTransition() -> PetState {
        let roll = Double.random(in: 0...1)
        if roll < 0.20 {
            return .looking
        }
        if roll < 0.38 {
            return .sleeping
        }
        if roll < 0.56 {
            return .stretching
        }
        if roll < 0.72 {
            return .meowing
        }
        return .walking
    }

    private func transition(to newState: PetState, at time: TimeInterval) {
        state = newState
        stateStartedAt = time

        switch newState {
        case .walking:
            nextTransitionAt = time + Double.random(in: 4...9)
        case .idle:
            nextTransitionAt = time + Double.random(in: 1.5...4)
        case .looking:
            nextTransitionAt = time + Double.random(in: 2...4)
        case .watching:
            nextTransitionAt = .infinity
        case .sleeping:
            nextTransitionAt = time + Double.random(in: 7...14)
        case .stretching:
            nextTransitionAt = time + 2.4
        case .meowing:
            nextTransitionAt = time + 2.0
        case .paused:
            nextTransitionAt = .infinity
        }
    }
}

File: Windows/PetCompanion.Windows/Pet/PetStateMachine.cs
namespace PetCompanion.Windows.Pet;

public enum PetState
{
    Walking,
    Idle,
    Looking,
    Watching,
    Sleeping,
    Rolling,
    Meowing,
    Paused
}

public sealed class PetStateMachine
{
    private readonly Random random = new();
    private PetState previousActiveState = PetState.Walking;
    private double nextTransitionAt = 6;

    public PetState State { get; private set; } = PetState.Walking;
    public bool IsPaused => State == PetState.Paused;
    public bool IsSleeping => State == PetState.Sleeping;

    public void SetPaused(bool paused, double time)
    {
        if (paused)
        {
            if (State != PetState.Paused)
            {
                previousActiveState = State;
                TransitionTo(PetState.Paused, time);
            }
        }
        else if (State == PetState.Paused)
        {
            TransitionTo(previousActiveState, time);
        }
    }

    public void Update(double time)
    {
        if (State == PetState.Paused || time < nextTransitionAt)
        {
            return;
        }

        switch (State)
        {
            case PetState.Walking:
                TransitionTo(PetState.Idle, time);
                break;
            case PetState.Idle:
                TransitionTo(NextIdleTransition(), time);
                break;
            case PetState.Looking:
            case PetState.Sleeping:
            case PetState.Rolling:
            case PetState.Meowing:
                TransitionTo(PetState.Walking, time);
                break;
            case PetState.Watching:
                nextTransitionAt = double.PositiveInfinity;
                break;
        }
    }

    public void SetWatching(bool watching, double time)
    {
        if (State == PetState.Paused)
        {
            return;
        }

        if (watching)
        {
            if (State is PetState.Walking or PetState.Idle or PetState.Looking)
            {
                TransitionTo(PetState.Watching, time);
            }
        }
        else if (State == PetState.Watching)
        {
            TransitionTo(PetState.Walking, time);
        }
    }

    public void TriggerClickReaction(double time)
    {
        if (State == PetState.Paused)
        {
            return;
        }

        TransitionTo(random.NextDouble() < 0.5 ? PetState.Meowing : PetState.Rolling, time);
    }

    public void TriggerSleep(double time)
    {
        if (State == PetState.Paused)
        {
            return;
        }

        TransitionTo(State == PetState.Sleeping ? PetState.Walking : PetState.Sleeping, time);
    }

    private PetState NextIdleTransition()
    {
        var roll = random.NextDouble();
        if (roll < 0.20)
        {
            return PetState.Looking;
        }
        if (roll < 0.38)
        {
            return PetState.Sleeping;
        }
        if (roll < 0.56)
        {
            return PetState.Rolling;
        }
        if (roll < 0.72)
        {
            return PetState.Meowing;
        }
        return PetState.Walking;
    }

    private void TransitionTo(PetState state, double time)
    {
        State = state;
        nextTransitionAt = state switch
        {
            PetState.Walking => time + RandomRange(4, 9),
            PetState.Idle => time + RandomRange(1.5, 4),
            PetState.Looking => time + RandomRange(2, 4),
            PetState.Watching => double.PositiveInfinity,
            PetState.Sleeping => time + RandomRange(7, 14),
            PetState.Rolling => time + 2.4,
            PetState.Meowing => time + 2.0,
            PetState.Paused => double.PositiveInfinity,
            _ => time + 3
        };
    }

    private double RandomRange(double min, double max) => min + random.NextDouble() * (max - min);
}
